Expose event information in on_save_position

dentarg opened the pull request:
This first commit is a somewhat quick and naive change, to have something to discuss.

Related to #23.

dentarg committed:
Pass the event type name to on_save_position

A somewhat quick and naive change, to have something to discuss.

Related to #23.

dentarg commented:
Now we have a of bit duplication between `RowEventListener` and `SaveEventListener` (the mapping of the name of some event types). One idea is to make an helper class for this, with [`row_type_to_string`](https://github.com/twingly/ecco/blob/v0.6.1/lib/ecco/row_event_listener.rb#L47-L55) and the different lists (`ROW_EVENTS`, `SAVE_EVENTS`), that the listeners can use. WDYT?

jage commented on lib/ecco/save_event_listener.rb:
```diff
@@ -2,30 +2,31 @@
         position = @client.get_binlog_position
 
-        @callback.call(filename, position)
+        @callback.call(event_type_name, filename, position)
```
I think `event_type_name` should be the last argument, since
- It's the least important, the main reason for this event is to persist the position, thus filename and position is more important.
- Keeps compatibility with old code: `Proc.new { |filename, position| # Will still work as before }`.
- Most likely to be ignore `Proc.new { |filename, position, _| # Save on all events }`.

dentarg replied:
Sure

jage commented on lib/ecco/save_event_listener.rb:
```diff
@@ -2,30 +2,31 @@
 
       case type
       when *accepted_events
+        event_type_name = SAVE_EVENTS[type]
```
`.fetch`, this should always work, but if it does not `fetch` will give a better stack trace.

dentarg replied:
Sure

dentarg committed:
Update spec TestHelper with event type name

Not sure if/how to make any assertions on event_type_name here.

jage reviewed (commented):
Sure, we should avoid duplication. But I think that can be fixed in a separate refactoring-PR once this has been verified to fix our issues.

dentarg committed:
The failure mode is better with Hash#fetch

dentarg committed:
Add event_type_name as the last argument

* It's the least important, the main reason for this event is to persist
  the position, thus filename and position is more important.

* Keeps compatibility with old code:

    Proc.new { |filename, position| # Will still work as before }

* Most likely to be ignored:

    Proc.new { |filename, position, _| # Save on all events }

jage commented on lib/ecco/save_event_listener.rb:
```diff
@@ -2,30 +2,31 @@
 
       case type
       when *accepted_events
+        event_type_name = SAVE_EVENTS.fetch(type)
```
How about moving this two rows down? Same order as the arguments.

dentarg replied:
Ah, yes, I forgot that.

jage replied:
Same thing in the spec.

dentarg replied:
Yes

jage commented:
I think this will work. Haven't looked at the specs, but I think the implementation is good.

dentarg committed:
Order and indent variables

dentarg committed:
Use correct order in the spec file too

dentarg commented:
Let's release this as 0.7.0, right?

jage commented on spec/lib/ecco/save_event_listener_spec.rb:
```diff
@@ -19,15 +19,18 @@
 
-      subject.callback = Proc.new do |filename, position|
+      subject.callback = Proc.new do |filename, position, event_type_name|
+        acutal_event_type_name = event_type_name
```
Move two rows down.

dentarg replied:
Done

jage commented on spec/lib/ecco/save_event_listener_spec.rb:
```diff
@@ -19,15 +19,18 @@
 
       subject.on_event(event)
 
+      expect(acutal_event_type_name).to eq("ROTATE")
```
Move two rows down.

dentarg replied:
Done

jage commented:
> Let's release this as 0.7.0, right?

Yes.

dentarg committed:
More ordering
